ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -1368,6 +1491,81 @@ private Offset getLastRecordOffset(LogSegment segment) {
     return lastOffset;
   }
 
+  // hardDeletePauseAndResume() helpers
+
+  /**
+   * Utility to test hard delete pause and resume with or without reload between pause and resume
+   * @param reloadIndex {@code true} if index has to be reloaded inbetween pause and resume. {@code false} otherwise
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws StoreException
+   */
+  private void testHardDeletePauseResume(boolean reloadIndex) throws InterruptedException, IOException, StoreException {
+    properties.put("store.deleted.message.retention.days", Integer.toString(1));
+    properties.put("store.hard.delete.bytes.per.sec", Integer.toString(Integer.MAX_VALUE / 10));
+    properties.setProperty("store.data.flush.interval.seconds", "3600");
+    // enable daemon thread to run hard deletes
+    properties.put("store.enable.hard.delete", "true");
+    reloadIndex(false);
+    assertTrue("Hard delete is not enabled", index.hardDeleter.isRunning());
+    // IndexSegment still uses real time so advance time so that it goes 2 days past the real time.
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    long expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // give it some time so that hard delete completes one cycle
+    waitUntilExpectedProgress(expectedProgress, 5000);
+    verifyEntriesForHardDeletes(deletedKeys);
+
+    Set<MockId> idsToDelete = new HashSet<>();
+    // delete two entries
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    Set<MockId> newIdsToDelete = new HashSet<>();
+    newIdsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    newIdsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : newIdsToDelete) {
+      addDeleteEntry(id);
+    }
```

Comment:
There are 3 instances at which deletes could occur. 
1. Deleted but not yet eligible to be hard deleted and pause is called before it could become hard deleted
2. After pause is called, new deletes which are eligible to be hard deleted
3. After reloading of index, new deletes that are eligible to be hard deleted
The test checks that all these entries are hard deleted after resume().